README.md
# tag-and-release-action

GitHub Action running a python script inside a docker container.

When the github-action detects a commit starting with `chg: :bookmark:`, and containing the text `preparing v`, it will automatically tag it, and create a release.

It will parse a CHANGELOG.md file with a format such as:

```markdown
# Changelog

## v1.0.1 (xxxx-xx-xx)

### Changes
...

## v1.0.0 (xxxx-xx-xx)

### Changes
...

```

And include it in the release's body.

That way, to release a new version of your application, just prepare your changelog accordingly, and commit with a message such as `chg: :bookmark: preparing v1.0.1`. Check this repository's releases and CHANGELOG.md as an example.

## Usage as a GitHub action

In your .github/workflows directory, create an `auto-tag-and-release.yml` file, and copy/paste the following code:

```yaml
name: auto-tag-and-release

on:
  push:
    branches: [ "dev" ]


  workflow_dispatch:

jobs:
  tag-and-release:
    uses: thentges/tag-and-release-action/.github/workflows/auto-tag-and-release.yml@v2

```


## Local usage

### Installation

- Create a GitHub API token here: https://github.com/settings/tokens/new with all `repo` scopes.
- Clone this repository on your machine
- Copy the file `local-entrypoint.py` to a new `auto-tag-and-release.py` file
- Change the permissions of this newly created file
```bash
sudo chmod 777 src/auto-tag-and-release.py
sudo chmod +x  src/auto-tag-and-release.py
```
- In this file, replace the first shebang line, with the path to a working python >= 3.8 environment with requests installed
- Add the path to this repository to your PATH
```bash
export PATH=$PATH:<path-to-repo-on-your-machine>
```
- Add the following environment variables to your .bashrc, or every-time you want to run the command
```bash
export INPUT_API_ACCESS_TOKEN=<the token created before cloning the repository>
export CHANGELOG_PATH=CHANGELOG.md
export LOCAL_REPO_OWNER=<name of the owner of the repository you are working in> # can be overridden anytime
```
- Add the following function to your .bashrc or .zshrc
```bash
git() {
    if [[ $@ == "auto-release" ]] ; then
        command auto-tag-and-release.py
    elif [[ $@ == "auto-release --tag" ]]; then
        command auto-tag-and-release.py --tag
    else
        command git "$@"
    fi
}
```

### Usage

Now that everything is installed you have access to the `git auto-release` command.

By default, git auto-release will require the tag to exist on your remote repository, and just parse your local CHANGELOG.md file to create a release.
You can change this behavior by passing the `--tag` flag, that will also create a tag when running the command if it does not exist yet.
If the release already exists, the program will exit, throwing an error.



CHANGELOG.md
# Changelog

## v2 (2022-07-15)

### Changes

* :children_crossing: use GITHUB_TOKEN instead of user-defined secret

## v1 (2022-07-15)

### New

* :sparkles: support for local usage without github-actions
* :tada: first version of tag-and-release-action


src/utils.py
import os
import sys
from typing import Tuple

from src.cfg import CHANGELOG_PATH, LOCAL_REPO_OWNER, ONLINE_REPO_NAME
from src.release_builder import GithubReleaseBuilder


def get_local_repository_name():
    # os.getcwd() return the absolute path of the current working directory
    # e.g: the directory from which this script is invoked
    # we take the last part of this part to get the directory name.
    return f"{LOCAL_REPO_OWNER}/{os.getcwd().split('/')[-1]}"


def parse_changelog() -> Tuple[str, str, str]:
    release_title = ""
    version = ""
    changelog = ""

    with open(CHANGELOG_PATH) as f:
        while line := f.readline():
            # first time we encounter H2
            # -> CURRENT release title
            if line.startswith("## ") and not release_title:
                release_title = line[3:]  # remove .md formatting
                version = release_title.split(" ")[0]  # take only the version
                # skip it from the body, since it is in the title
                continue

            # second time we encounter H2
            # -> PREVIOUS release title
            elif line.startswith("## "):
                break

            # skip all H1 headers, the only H1 should be "Changelog"
            elif line.startswith("# "):
                continue

            changelog += line

    if not release_title or not version or not changelog:
        raise Exception("Could not parse changelog")

    return release_title, version, changelog


def execute(auto_tag: bool = False):
    release_title, version, changelog = parse_changelog()

    release_builder = GithubReleaseBuilder(
        REPO_NAME=ONLINE_REPO_NAME or get_local_repository_name(),
        TAG=version,
        TITLE=release_title,
        CHANGELOG=changelog,
    )

    # in case the tag does not exist already, do not create release to avoid creating
    # wrong releases in case the version parsing failed for any reason (wrong changelog format, ...)
    # just pass auto_tag=True if you want to remove this security, which is a bit riskier.
    # note that it will check among the last 15 tags only, because of the per_page setting we make.
    if not auto_tag and not release_builder.already_tagged:
        print(
            f"Tag {release_builder.TAG} does not exist on {release_builder.remote_url}\n"
            f"Please make sure you created the tag and reviewed your changelog."
        )
        exit(1)

    if release_builder.already_exists:
        print(
            f"Release for tag {release_builder.TAG} already exists on {release_builder.remote_url}\n"
            f"Please review your changelog."
        )
        exit(1)

    # ##### 3. creation
    release_builder.create()
    print(
        f"To github.com:{release_builder.remote_url}\n"
        f"  * [new release]         {release_builder.TITLE}",
        end="",
    )


def parse_autotag_flag() -> bool:
    try:
        flag = sys.argv[1]
    except IndexError:
        return False

    return flag == "--tag"


src/release_builder.py
from dataclasses import dataclass
from functools import cached_property
from typing import Union

import requests
from requests import Response

from src.cfg import GITHUB_ACCESS_TOKEN


@dataclass
class GithubReleaseBuilder:
    REPO_NAME: str
    TAG: str
    TITLE: str
    CHANGELOG: str

    @property
    def remote_url(self) -> str:
        return f"{self.REPO_NAME}.git"

    @property
    def headers(self):
        return {"Authorization": f"token {GITHUB_ACCESS_TOKEN}"}

    @cached_property
    def already_tagged(self) -> bool:
        res = self._make_github_request("/tags?per_page=15", method="get")
        return self.TAG in [tag["name"] for tag in res.json()]

    @cached_property
    def already_exists(self) -> bool:
        res = self._make_github_request(f"/releases/tags/{self.TAG}", method="get")
        if res.status_code == 404:
            return False

        return True

    def create(self) -> None:
        res = self._make_github_request(
            f"/releases",
            json={
                "tag_name": self.TAG,
                "name": self.TITLE,
                "body": self.CHANGELOG,
            },
            method="post",
        )

        if not res.ok:
            print(_pretty_github_errors_str(res, release_title=self.TITLE))
            exit(1)

    def _make_github_request(self, endpoint, method: str, *args, **kwargs) -> Response:
        return getattr(requests, method)(
            f"https://api.github.com/repos/{self.REPO_NAME}{endpoint}",
            headers=self.headers,
            *args,
            **kwargs,
        )


def _pretty_github_errors_str(
    response: Response, release_title: str
) -> Union[str, dict]:
    try:
        _lines = "\n".join(
            f"  * {err['field']}: {err['code']}" for err in response.json()["errors"]
        )
        return f"Could not create release: {release_title}" + _lines
    except:
        return response.json()


script.py
#!/Users/Thibault/Code/random-tests/venv/bin/python

import os
import sys
from dataclasses import dataclass
from functools import cached_property
from typing import Tuple, Union

import requests
from requests import Response

CHANGELOG_PATH = f"github-releases/MOCK_CHANGELOG.md"
GITHUB_USERNAME = "thentges"


def get_github_access_token() -> str:
    return os.getenv("API_ACCESS_TOKEN")


def parse_changelog(changelog_filepath: str = CHANGELOG_PATH) -> Tuple[str, str, str]:
    release_title = ""
    version = ""
    changelog = ""

    with open(changelog_filepath) as f:
        while line := f.readline():
            # first time we encounter H2
            # -> CURRENT release title
            if line.startswith("## ") and not release_title:
                release_title = line[3:]  # remove .md formatting
                version = release_title.split(" ")[0]  # take only the version
                # skip it from the body, since it is in the title
                continue

            # second time we encounter H2
            # -> PREVIOUS release title
            elif line.startswith("## "):
                break

            # skip all H1 headers, the only H1 should be "Changelog"
            elif line.startswith("# "):
                continue

            changelog += line

    if not release_title or not version or not changelog:
        raise Exception("Could not parse changelog")

    return release_title, version, changelog


@dataclass
class GithubReleaseBuilder:
    REPO_NAME: str
    TAG: str
    TITLE: str
    CHANGELOG: str
    ACCESS_TOKEN: str

    USERNAME: str = GITHUB_USERNAME

    @property
    def full_repo(self) -> str:
        return f"{self.USERNAME}/{self.REPO_NAME}"

    @property
    def remote_url(self) -> str:
        return f"{self.full_repo}.git"

    @property
    def headers(self):
        return {"Authorization": f"token {self.ACCESS_TOKEN}"}

    @cached_property
    def already_tagged(self) -> bool:
        res = self._make_github_request("/tags?per_page=15", method="get")
        return self.TAG in [tag["name"] for tag in res.json()]

    @cached_property
    def already_exists(self) -> bool:
        res = self._make_github_request(f"/releases/tags/{self.TAG}", method="get")
        if res.status_code == 404:
            return False

        return True

    def create(self) -> None:
        res = self._make_github_request(
            f"/releases",
            json={
                "tag_name": self.TAG,
                "name": self.TITLE,
                "body": self.CHANGELOG,
            },
            method="post",
        )

        if not res.ok:
            print(_pretty_github_errors_str(res, release_title=self.TITLE))
            exit(1)

    def _make_github_request(self, endpoint, method: str, *args, **kwargs) -> Response:
        return getattr(requests, method)(
            f"https://api.github.com/repos/{self.full_repo}{endpoint}",
            headers=self.headers,
            *args,
            **kwargs,
        )


def _pretty_github_errors_str(
    response: Response, release_title: str
) -> Union[str, dict]:
    try:
        _lines = "\n".join(
            f"  * {err['field']}: {err['code']}" for err in response.json()["errors"]
        )
        return f"Could not create release: {release_title}" + _lines
    except:
        return response.json()


def execute(auto_tag: bool = False):
    # ##### 1. parse environment
    access_token = get_github_access_token()
    if not access_token:
        print(
            "No access token found.\n"
            "Make sure you set the API_ACCESS_TOKEN environment variable."
        )
        exit(1)

    release_title, version, changelog = parse_changelog()

    # prepare the factory
    release_builder = GithubReleaseBuilder(
        # os.getcwd() return the absolute path of the current working directory
        # e.g: the directory from which this script is invoked
        # we take the last part of this part to get the directory name
        REPO_NAME=os.getcwd().split("/")[-1],
        TAG=version,
        TITLE=release_title,
        CHANGELOG=changelog,
        ACCESS_TOKEN=access_token,
    )

    # ##### 2. validation

    # in case the tag does not exist already, do not create release to avoid creating
    # wrong releases in case the version parsing failed for any reason (wrong changelog format, ...)
    # just pass auto_tag=True if you want to remove this security, which is a bit riskier.
    # note that it will check among the last 15 tags only, because of the per_page setting we make.
    if not auto_tag and not release_builder.already_tagged:
        print(
            f"Tag {release_builder.TAG} does not exist on {release_builder.remote_url}\n"
            f"Please make sure you created the tag and reviewed your changelog."
        )
        exit(1)

    if release_builder.already_exists:
        print(
            f"Release for tag {release_builder.TAG} already exists on {release_builder.remote_url}\n"
            f"Please review your changelog."
        )
        exit(1)

    # ##### 3. creation
    release_builder.create()
    print(
        f"To github.com:{release_builder.remote_url}\n"
        f"  * [new release]         {release_builder.TITLE}",
        end="",
    )


def parse_autotag_flag() -> bool:
    try:
        flag = sys.argv[1]
    except IndexError:
        return False

    return flag == "--tag"


execute(auto_tag=parse_autotag_flag())
